Small rotating drum example (#519)
lethe-cfd/lethe: doc/source/examples/dem/small-scale-rotating-drum/small-scale-rotating-drum.rst

==================================
Small Scale Rotating Drum (3d)
==================================

> This example of Lethe-DEM is to simulate the behaviour of dry granular flow in a small scale rotating drum. The discrete element method DEM will be responsible of describing the behaviour of particles. DEM parameters are described in Lethe-DEM documentation, i.e. `DEM parameters <../../../parameters/dem/dem.html>`_ for more detailed information.

Reviewer: ```suggestion
This Lethe-DEM example simulates dry granular flow behavior in a small-scale rotating drum. The discrete element method DEM will be responsible for describing the behavior of particles. DEM parameters are described in Lethe-DEM documentation, i.e. `DEM parameters <../../../parameters/dem/dem.html>`_ for more detailed information.
```

Features
----------------------------------
- Solvers: ``dem_3d``
- Three-dimensional problem
- Rotational boundary
- Load-balancing

Location of the examples
------------------------
 ``/examples/dem/small_scale_rotating_drum/small_rotating_drum_dem.prm``

Description of the case
-----------------------

> This example simulates a cascading regime in a small scale rotating drum. First, we first use Lethe-DEM to fill the bed with 20000 particles. We enable check-pointing in order to write the DEM checkpoint files which will be used as the starting point of the DEM simulations . Then, the solver ``dem_3d`` is used to simulate the behaviour of dry granular flow within the rotating drum.

Reviewer 2: ```suggestion
This example simulates a cascading regime in a small scale rotating drum. First, we use Lethe-DEM to fill the bed with 20000 particles. We enable check-pointing in order to write the DEM checkpoint files which will be used as the starting point of the DEM simulations. Then, the solver ``dem_3d`` is used to simulate the behaviour of dry granular flow within the rotating drum.
```

> This example simulates a cascading regime in a small scale rotating drum. First, we first use Lethe-DEM to fill the bed with 20000 particles. We enable check-pointing in order to write the DEM checkpoint files which will be used as the starting point of the DEM simulations . Then, the solver ``dem_3d`` is used to simulate the behaviour of dry granular flow within the rotating drum.

Reviewer 2: I don't understand this sentence: 
`We enable check-pointing in order to write the DEM checkpoint files which will be used as the starting point of the DEM simulations .`
You enable checkpointing  to be able to restart the DEM simulation? Or do you somehow do a simulation before to obtain files that are used to start a new simulation?

Parameter file
--------------

Mesh
~~~~~

> In this example, we choose a ``cylinder`` grid type to create a cylinder. Grid arguments are the radius of the cylinder (0.056 m) and half-length (0.051m) , respectively.  The grid is refined 3 times. The ``expand particle-wall contact search`` is used in concave geometries to enable extended particle-wall contact search with boundary faces of neighbor cells for particles located in each boundary cell (for more details  see `Rotating Drum example <../rotating-drum/rotating-drum.html>`_).

Reviewer: ```suggestion
In this example, we choose a ``cylinder`` grid type to create a cylinder. Grid arguments are the radius of the cylinder (0.056 m) and half-length (0.051m), respectively.  The grid is refined 3 times, using ``set initial refinement``. The ``expand particle-wall contact search`` is used in concave geometries to enable extended particle-wall contact search with boundary faces of neighbor cells for particles located in each boundary cell (for more details see `Rotating Drum example <../rotating-drum/rotating-drum.html>`_).
```

.. code-block:: text

    subsection mesh
        set type                 				= dealii
        set grid type      	     				= cylinder
        set grid arguments       				= 0.056:0.051
        set initial refinement   				= 3
    end

Packing information
~~~~~~~~~~~~~~~~~~~~

An insertion box is defined inside the cylindrical domain, inserting 8000 particles every 0.5 seconds while the cylinder is at rest.

.. code-block:: text

    subsection insertion info
    	set insertion method				        = non_uniform
    	set inserted number of particles at each time step      = 8000
    	set insertion frequency            		 	= 100000
    	set insertion box minimum x            	 	= -0.05
    	set insertion box minimum y            	        = 0.
    	set insertion box minimum z            	        = -0.04
    	set insertion box maximum x            	        = 0.05
    	set insertion box maximum y           	 	= 0.04
    	set insertion box maximum z            	        = 0.04
    	set insertion distance threshold			= 1.1
    	set insertion random number range			= 0.05
    	set insertion random number seed			= 19
    end

> Restart files are written once the packing ends.

Reviewer: You could explain here what the Restart files are used for.

Lagrangian physical properties
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The particles are monodispersed with a radius 0.0015 m  and a density of 2500 kg/m3, respectively.

Reviewer: I would explain how to do it. There are many parameters in this section with no explanation provided.

.. code-block:: text

    subsection lagrangian physical properties
        set gx            		 						= 0.0
        set gy            		 						= -9.81
        set gz            		 						= 0.0
        set number of particle types	                = 1
            subsection particle type 0
            set size distribution type					= uniform
                set diameter            	 			= 0.003
                set number              				= 20000
                set density particles  	 				= 2500
                set young modulus particles         	= 100000000
                set poisson ratio particles          	= 0.24
                set restitution coefficient particles	= 0.97
                set friction coefficient particles      = 0.3
                set rolling friction particles         	= 0.01
        end
        set young modulus wall            				= 100000000
        set poisson ratio wall            				= 0.24
        set restitution coefficient wall           		= 0.85
        set friction coefficient wall         			= 0.35

>         set rolling friction wall         	      	  	= 0.01

Reviewer 2: Do you have any further comments on these parameters?

    end

Model parameters
~~~~~~~~~~~~~~~~~

> The load balancing method used is the dynamic method, more details are explained in the `Rotating Drum example <../rotating-drum/rotating-drum.html>`_. 

Reviewer: Here you have more parameters than the load balancing-related ones.

Reviewer: If you want to focus on the load balancing part, an extra period explaining it would be nice :)

> The load balancing method used is the dynamic method, more details are explained in the `Rotating Drum example <../rotating-drum/rotating-drum.html>`_. 

Reviewer 3: ```suggestion
The load balancing method used is the dynamic method. This method checks frequently if load balancing should be applied based on a user inputted frequency.  Load balancing is dynamically applied if a certain condition is applied. More details regarding load balancing are explained in the `Rotating Drum example <../rotating-drum/rotating-drum.html>`_. 
```

    subsection model parameters
      set contact detection method 		   	 = dynamic
      set dynamic contact search size coefficient	 = 0.8
      set neighborhood threshold			 = 1.3
      set load balance method				 = dynamic
      set load balance threshold			 = 0.5
      set dynamic load balance check frequency		 = 100000
      set particle particle contact force method         = hertz_mindlin_limit_overlap
      set particle wall contact force method             = nonlinear
      set integration method				 = velocity_verlet
    end

Boundary Condition
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The rotation of the cylinder is applied using a rotational boundary condition with a value of 8 rad/s over the x axis.

.. code-block:: text

    subsection DEM boundary conditions
      set number of boundary conditions         = 1
        subsection boundary condition 0
            set boundary id					= 0
            set type              				= rotational
            set rotational speed				= 8
            set rotational vector x				= 1
            set rotational vector y				= 0
            set rotational vector z				= 0
        end
    end

Simulation control
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The simulation runs for 12 seconds of real time.

Reviewer 2: How do you choose this 12 seconds? Is there any reason behind this?

> The simulation runs for 12 seconds of real time.

Reviewer 3: ```suggestion
The end time of the simulation is 10 seconds in real time. We set the time equal to 12 seconds as the simulation is restarted after the packing dem simulation which was run for 2 seconds in real time.
```

.. code-block:: text

    subsection simulation control
      set time step                 		 = 5e-6
      set time end       			 = 12
      set log frequency				 = 2000
      set output frequency            		 = 2000
      set output path                  	 	 = ./output_dem/
    end

Running the simulation
----------------------
The simulation is launched on two steps:
 
For packing: 

.. code-block:: text

   mpirun -np 6 dem_3d packing_rotating_drum.prm

For cylinder rotation: 

.. code-block:: text
 
   mpirun -np 6 dem_3d small_rotating_drum_dem.prm

.. warning::

> 	This example needs a simulation time of approximately 6 hours 6 cores.s.

Reviewer 2: I think it would be useful to have some more details here. 6 hours, 6 cores, in what kind of machine?  And what is 6 cores.s?

Results
---------

> Velocity profile of particles during the simulation:

Reviewer 2: Nice image. Would you like to explain a little bit more? Does it make sense? Why is the velocity profile like this?

                set restitution coefficient particles	= 0.97
                set friction coefficient particles      = 0.3
                set rolling friction particles         	= 0.01
        end
        set young modulus wall            				= 100000000
        set poisson ratio wall            				= 0.24
        set restitution coefficient wall           		= 0.85
        set friction coefficient wall         			= 0.35
        set rolling friction wall         	      	  	= 0.01
    end

Model parameters
~~~~~~~~~~~~~~~~~

The load balancing method used is the dynamic method. This method checks frequently if load balancing should be applied based on a user inputted frequency. Load balancing is dynamically applied if a certain condition is applied. More details regarding load balancing are explained in the `Rotating Drum example <../rotating-drum/rotating-drum.html>`_. 

.. code-block:: text

    subsection model parameters
      set contact detection method 		   	 = dynamic
      set dynamic contact search size coefficient	 = 0.8
      set neighborhood threshold			 = 1.3
      set load balance method				 = dynamic
      set load balance threshold			 = 0.5

>       set dynamic load balance check frequency		 = 100000

Reviewer 4: You can put a lower frequency than that. 100k is not frequent enough. on my machine I tried with 10k and it was working well.

        end
    end

Simulation control
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The end time of the simulation is 10 seconds in real time. We set the time equal to 12 seconds as the simulation is restarted after the packing dem simulation which was run for 2 seconds in real time.

.. code-block:: text

    subsection simulation control
      set time step                 		 = 5e-6
      set time end       			 = 12
      set log frequency				 = 2000
      set output frequency            		 = 2000
      set output path                  	 	 = ./output_dem/
    end

Running the simulation
-----------------------

The simulation is launched on two steps, the first step packs the particle in the cylinder. The second step rotates the drum and simulates the movement of the particles. The simulation could have been performed in a single step, however; doing it this way allows us to use the restart files of the packing to run the simulation using DEM for a dry simulation and a CFD-DEM for a wet simulation without having to pack the particles twice.

For packing:

.. code-block:: text

   mpirun -np 6 dem_3d packing_rotating_drum.prm
   mpirun -np 6 dem_3d small_rotating_drum_dem.prm

.. warning::

> 	This example needs a simulation time of approximately 6 hours on 6 cores.s.

Reviewer 2: I still think that we need to specify on what machine. What is the meaning of 6 cores.s?

Results
---------

> Velocity profile of particles during the simulation:

Reviewer 4: Would you like to add some thing where you change the velocity to a higher or lower value and demonstrate another flow regime?

> .. image:: image/rotating_drum.png

Reviewer 4: There are also other things you could analyse. 
- Make a movie of the results instead of a static image. that would be fun
- Make a few images of the processor distribution in the domain for a few times, that would be cool to show
:)
